ICE: const eval: `expected type differs from actual type.` `()` vs `i32`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code

```Rust
fn main() {
    [(); {
        while true {
            break 2;
        }
        201
    }];
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
9565b68b741fd0466b3af075139e819e7d73ee38
```

### Error output

```
warning: denote infinite loops with `loop { ... }`
 --> ./847B579CB9D548BFF192F0D147D9628805BF639194C504BD0CB09AF68207F69F.rs:3:9
  |
3 |         while true {
  |         ^^^^^^^^^^ help: use `loop`
  |
  = note: `#[warn(while_true)]` on by default

error[E0571]: `break` with value from a `while` loop
 --> ./847B579CB9D548BFF192F0D147D9628805BF639194C504BD0CB09AF68207F69F.rs:4:13
  |
3 |         while true {
  |         ---------- you can't `break` with a value in a `while` loop
4 |             break 2;
  |             ^^^^^^^ can only break with a value inside `loop` or breakable block
  |
help: use `break` on its own without a value inside this `while` loop
  |
4 |             break;
  |             ~~~~~
```


<!-- TRIAGEBOT_START -->

<!-- TRIAGEBOT_ASSIGN_START -->

<!-- TRIAGEBOT_ASSIGN_DATA_START$${"user":"gurry"}$$TRIAGEBOT_ASSIGN_DATA_END -->

<!-- TRIAGEBOT_ASSIGN_END -->
<!-- TRIAGEBOT_END -->

````
 --> 114529.rs:3:9
  |
3 |         while true {
  |         ^^^^^^^^^^ help: use `loop`
  |
  = note: `#[warn(while_true)]` on by default

error[E0571]: `break` with value from a `while` loop
 --> 114529.rs:4:13
  |
3 |         while true {
  |         ---------- you can't `break` with a value in a `while` loop
4 |             break 2;
  |             ^^^^^^^ can only break with a value inside `loop` or breakable block
  |
help: use `break` on its own without a value inside this `while` loop
  |
4 |             break;
  |             ~~~~~

error: internal compiler error: compiler/rustc_const_eval/src/interpret/place.rs:836:13: type mismatch when copying!
                                src: i32,
                                dest: ()
 --> 114529.rs:4:19
  |
4 |             break 2;
  |                   ^

thread 'rustc' panicked at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/compiler/rustc_errors/src/lib.rs:995:33:
Box<dyn Any>
stack backtrace:
   0:     0x7f51cab62f0c - std::backtrace_rs::backtrace::libunwind::trace::h125123d4e246cce7
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f51cab62f0c - std::backtrace_rs::backtrace::trace_unsynchronized::h793606c4e4c86869
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f51cab62f0c - std::sys_common::backtrace::_print_fmt::h00383e358727df48
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/sys_common/backtrace.rs:67:5
   3:     0x7f51cab62f0c - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h8fb6045ca0224bc5
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f51cabc98bc - core::fmt::rt::Argument::fmt::h86e183b82477767e
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/core/src/fmt/rt.rs:138:9
   5:     0x7f51cabc98bc - core::fmt::write::h41479ddc3ae3bd77
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/core/src/fmt/mod.rs:1116:21
   6:     0x7f51cab55fbf - std::io::Write::write_fmt::h62d2da41ce39b458
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/io/mod.rs:1763:15
   7:     0x7f51cab62cf4 - std::sys_common::backtrace::_print::h7ca8c582824e0f2f
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x7f51cab62cf4 - std::sys_common::backtrace::print::ha89296afca8c7a71
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x7f51cab65b07 - std::panicking::default_hook::{{closure}}::h938b29914e951237
  10:     0x7f51cab6584a - std::panicking::default_hook::hbe391f4c4a062ffb
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/panicking.rs:292:9
  11:     0x7f51cd84e505 - std[77c63684c9947df9]::panicking::update_hook::<alloc[e669280e87f40af2]::boxed::Box<rustc_driver_impl[600953db8fbf49d8]::install_ice_hook::{closure#0}>>::{closure#0}
  12:     0x7f51cab662a8 - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::h54184735b80a55c7
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/alloc/src/boxed.rs:2021:9
  13:     0x7f51cab662a8 - std::panicking::rust_panic_with_hook::h0e8a4b840e3d8850
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/panicking.rs:735:13
  14:     0x7f51cd7e34f4 - std[77c63684c9947df9]::panicking::begin_panic::<rustc_errors[3edaeedde173e4ec]::ExplicitBug>::{closure#0}
  15:     0x7f51cd7d88c6 - std[77c63684c9947df9]::sys_common::backtrace::__rust_end_short_backtrace::<std[77c63684c9947df9]::panicking::begin_panic<rustc_errors[3edaeedde173e4ec]::ExplicitBug>::{closure#0}, !>
  16:     0x7f51cd7d4486 - std[77c63684c9947df9]::panicking::begin_panic::<rustc_errors[3edaeedde173e4ec]::ExplicitBug>
  17:     0x7f51cd7cb14e - <rustc_errors[3edaeedde173e4ec]::HandlerInner>::span_bug::<rustc_span[4a2fc62f15bce35e]::span_encoding::Span, alloc[e669280e87f40af2]::string::String>
  18:     0x7f51cd7c5aab - <rustc_errors[3edaeedde173e4ec]::Handler>::span_bug::<rustc_span[4a2fc62f15bce35e]::span_encoding::Span, alloc[e669280e87f40af2]::string::String>
  19:     0x7f51cd7e382b - rustc_middle[654db2abf46e4245]::util::bug::opt_span_bug_fmt::<rustc_span[4a2fc62f15bce35e]::span_encoding::Span>::{closure#0}
  20:     0x7f51cd7e385a - rustc_middle[654db2abf46e4245]::ty::context::tls::with_opt::<rustc_middle[654db2abf46e4245]::util::bug::opt_span_bug_fmt<rustc_span[4a2fc62f15bce35e]::span_encoding::Span>::{closure#0}, !>::{closure#0}
  21:     0x7f51cd7da068 - rustc_middle[654db2abf46e4245]::ty::context::tls::with_context_opt::<rustc_middle[654db2abf46e4245]::ty::context::tls::with_opt<rustc_middle[654db2abf46e4245]::util::bug::opt_span_bug_fmt<rustc_span[4a2fc62f15bce35e]::span_encoding::Span>::{closure#0}, !>::{closure#0}, !>
  22:     0x7f51ccca2864 - rustc_middle[654db2abf46e4245]::util::bug::span_bug_fmt::<rustc_span[4a2fc62f15bce35e]::span_encoding::Span>
  23:     0x7f51cd1d5659 - rustc_const_eval[d6e1d76b0b47622a]::const_eval::eval_queries::eval_body_using_ecx
  24:     0x7f51ccc35012 - rustc_const_eval[d6e1d76b0b47622a]::const_eval::eval_queries::eval_to_allocation_raw_provider
  25:     0x7f51cc2ecc56 - rustc_query_impl[a1c0fd0e19622223]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>
  26:     0x7f51cc2ec836 - rustc_query_system[77a688ff6f07ed38]::query::plumbing::try_execute_query::<rustc_query_impl[a1c0fd0e19622223]::DynamicConfig<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[a1c0fd0e19622223]::plumbing::QueryCtxt, false>
  27:     0x7f51cd319774 - rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  28:     0x7f51ccc39ba6 - rustc_middle[654db2abf46e4245]::query::plumbing::query_get_at::<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>>
  29:     0x7f51ccc356ba - rustc_const_eval[d6e1d76b0b47622a]::const_eval::eval_queries::eval_to_allocation_raw_provider
  30:     0x7f51cc2ecc56 - rustc_query_impl[a1c0fd0e19622223]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_allocation_raw::dynamic_query::{closure#2}::{closure#0}, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>
  31:     0x7f51cc2ec836 - rustc_query_system[77a688ff6f07ed38]::query::plumbing::try_execute_query::<rustc_query_impl[a1c0fd0e19622223]::DynamicConfig<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[a1c0fd0e19622223]::plumbing::QueryCtxt, false>
  32:     0x7f51cd319774 - rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_allocation_raw::get_query_non_incr::__rust_end_short_backtrace
  33:     0x7f51ccc39ba6 - rustc_middle[654db2abf46e4245]::query::plumbing::query_get_at::<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>>
  34:     0x7f51cd0c537f - <rustc_const_eval[d6e1d76b0b47622a]::provide::{closure#0} as core[158b56b61cd52734]::ops::function::FnOnce<(rustc_middle[654db2abf46e4245]::ty::context::TyCtxt, rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>)>>::call_once
  35:     0x7f51cc2ed5c6 - rustc_query_impl[a1c0fd0e19622223]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_valtree::dynamic_query::{closure#2}::{closure#0}, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>
  36:     0x7f51cc2ed57f - <rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_valtree::dynamic_query::{closure#2} as core[158b56b61cd52734]::ops::function::FnOnce<(rustc_middle[654db2abf46e4245]::ty::context::TyCtxt, rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>)>>::call_once
  37:     0x7f51cc2ec8b0 - rustc_query_system[77a688ff6f07ed38]::query::plumbing::try_execute_query::<rustc_query_impl[a1c0fd0e19622223]::DynamicConfig<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>, false, false, false>, rustc_query_impl[a1c0fd0e19622223]::plumbing::QueryCtxt, false>
  38:     0x7f51cd319b54 - rustc_query_impl[a1c0fd0e19622223]::query_impl::eval_to_valtree::get_query_non_incr::__rust_end_short_backtrace
  39:     0x7f51cc75f57d - rustc_middle[654db2abf46e4245]::query::plumbing::query_get_at::<rustc_query_system[77a688ff6f07ed38]::query::caches::DefaultCache<rustc_middle[654db2abf46e4245]::ty::ParamEnvAnd<rustc_middle[654db2abf46e4245]::mir::interpret::GlobalId>, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 24usize]>>>
  40:     0x7f51cc75ea24 - <rustc_middle[654db2abf46e4245]::ty::context::TyCtxt>::const_eval_global_id_for_typeck
  41:     0x7f51cc75dc04 - <rustc_middle[654db2abf46e4245]::ty::context::TyCtxt>::const_eval_resolve_for_typeck
  42:     0x7f51cc75b456 - <rustc_middle[654db2abf46e4245]::ty::consts::Const>::eval
  43:     0x7f51cc2a6044 - <rustc_trait_selection[12e12a1b98cf9c36]::traits::project::AssocTypeNormalizer as rustc_type_ir[a26b45915bcf3b66]::fold::TypeFolder<rustc_middle[654db2abf46e4245]::ty::context::TyCtxt>>::fold_const
  44:     0x7f51cd0292de - rustc_trait_selection[12e12a1b98cf9c36]::traits::project::normalize_with_depth_to::<rustc_middle[654db2abf46e4245]::ty::consts::Const>::{closure#0}
  45:     0x7f51cd028cd8 - <rustc_hir_typeck[6811c5555fb14140]::fn_ctxt::FnCtxt>::array_length_to_const
  46:     0x7f51cbfe3959 - <rustc_hir_typeck[6811c5555fb14140]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  47:     0x7f51cc064e72 - <rustc_hir_typeck[6811c5555fb14140]::fn_ctxt::FnCtxt>::check_block_with_expected
  48:     0x7f51cbfdf41a - <rustc_hir_typeck[6811c5555fb14140]::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  49:     0x7f51cbc28b67 - rustc_hir_typeck[6811c5555fb14140]::check::check_fn
  50:     0x7f51cbb735c3 - rustc_hir_typeck[6811c5555fb14140]::typeck
  51:     0x7f51cbd8d2f7 - rustc_query_impl[a1c0fd0e19622223]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[a1c0fd0e19622223]::query_impl::typeck::dynamic_query::{closure#2}::{closure#0}, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 8usize]>>
  52:     0x7f51cbd87162 - rustc_query_system[77a688ff6f07ed38]::query::plumbing::try_execute_query::<rustc_query_impl[a1c0fd0e19622223]::DynamicConfig<rustc_query_system[77a688ff6f07ed38]::query::caches::VecCache<rustc_span[4a2fc62f15bce35e]::def_id::LocalDefId, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 8usize]>>, false, false, false>, rustc_query_impl[a1c0fd0e19622223]::plumbing::QueryCtxt, false>
  53:     0x7f51cbd88684 - rustc_query_impl[a1c0fd0e19622223]::query_impl::typeck::get_query_non_incr::__rust_end_short_backtrace
  54:     0x7f51ccd915ee - rustc_hir_analysis[b86c677b537b6988]::check_crate
  55:     0x7f51ccd86b5a - rustc_interface[56f67b4ff5f3f802]::passes::analysis
  56:     0x7f51ccfe9c57 - rustc_query_impl[a1c0fd0e19622223]::plumbing::__rust_begin_short_backtrace::<rustc_query_impl[a1c0fd0e19622223]::query_impl::analysis::dynamic_query::{closure#2}::{closure#0}, rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 1usize]>>
  57:     0x7f51ccfe9846 - rustc_query_system[77a688ff6f07ed38]::query::plumbing::try_execute_query::<rustc_query_impl[a1c0fd0e19622223]::DynamicConfig<rustc_query_system[77a688ff6f07ed38]::query::caches::SingleCache<rustc_middle[654db2abf46e4245]::query::erase::Erased<[u8; 1usize]>>, false, false, false>, rustc_query_impl[a1c0fd0e19622223]::plumbing::QueryCtxt, false>
  58:     0x7f51ccfe95fd - rustc_query_impl[a1c0fd0e19622223]::query_impl::analysis::get_query_non_incr::__rust_end_short_backtrace
  59:     0x7f51ccf03dcd - std[77c63684c9947df9]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[56f67b4ff5f3f802]::util::run_in_thread_with_globals<rustc_interface[56f67b4ff5f3f802]::interface::run_compiler<core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>, rustc_driver_impl[600953db8fbf49d8]::run_compiler::{closure#1}>::{closure#0}, core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>>
  60:     0x7f51ccf03083 - <<std[77c63684c9947df9]::thread::Builder>::spawn_unchecked_<rustc_interface[56f67b4ff5f3f802]::util::run_in_thread_with_globals<rustc_interface[56f67b4ff5f3f802]::interface::run_compiler<core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>, rustc_driver_impl[600953db8fbf49d8]::run_compiler::{closure#1}>::{closure#0}, core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[158b56b61cd52734]::result::Result<(), rustc_span[4a2fc62f15bce35e]::ErrorGuaranteed>>::{closure#1} as core[158b56b61cd52734]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  61:     0x7f51cab71225 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hba13f6aa2e84f743
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/alloc/src/boxed.rs:2007:9
  62:     0x7f51cab71225 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h576a2cf3e706c4e2
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/alloc/src/boxed.rs:2007:9
  63:     0x7f51cab71225 - std::sys::unix::thread::Thread::new::thread_start::h98bac879f1e5e8c0
                               at /rustc/eb0f3ed59c6508a37c6598bc9762987f053993a7/library/std/src/sys/unix/thread.rs:108:17
  64:     0x7f51ca88c9eb - <unknown>
  65:     0x7f51ca910dfc - <unknown>
  66:                0x0 - <unknown>

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please attach the file at `/home/matthias/vcs/github/glacier2/fixed/rustc-ice-2023-10-03T13:52:29.879656559Z-2978437.txt` to your bug report

query stack during panic:
#0 [eval_to_allocation_raw] const-evaluating + checking `main::{constant#0}`
#1 [eval_to_allocation_raw] const-evaluating + checking `main::{constant#0}`
#2 [eval_to_valtree] evaluating type-level constant
#3 [typeck] type-checking `main`
#4 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 2 previous errors; 1 warning emitted

For more information about this error, try `rustc --explain E0571`.
````

--- a/compiler/rustc_hir_typeck/src/expr.rs
+++ b/compiler/rustc_hir_typeck/src/expr.rs
@@ -625,10 +625,15 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                     }
                 };
 
-                // If the loop context is not a `loop { }`, then break with
-                // a value is illegal, and `opt_coerce_to` will be `None`.
-                // Just set expectation to error in that case.
-                let coerce_to = opt_coerce_to.unwrap_or_else(|| Ty::new_misc_error(tcx));
+                let coerce_to = match opt_coerce_to {
+                    Some(c) => c,
+                    None => {
+                        // If the loop context is not a `loop { }`, then break with
+                        // a value is illegal, and `opt_coerce_to` will be `None`.
+                        // Return error in that case (#114529).
+                        return Ty::new_misc_error(tcx);
+                    }
+                };
 
                 // Recurse without `enclosing_breakables` borrowed.
                 e_ty = self.check_expr_with_hint(e, coerce_to);

--- a/tests/ui/typeck/issue-114529-illegal-break-with-value.rs
+++ b/tests/ui/typeck/issue-114529-illegal-break-with-value.rs
@@ -1,0 +1,20 @@
+// Regression test for issue #114529
+// Tests that we do not ICE during const eval for a
+// break-with-value in contexts where it is illegal
+
+#[allow(while_true)]
+fn main() {
+    [(); {
+        while true {
+            break 9; //~ ERROR `break` with value from a `while` loop
+        };
+        51
+    }];
+
+    [(); {
+        while let Some(v) = Some(9) {
+            break v; //~ ERROR `break` with value from a `while` loop
+        };
+        51
+    }];
+}

--- a/tests/ui/typeck/issue-114529-illegal-break-with-value.stderr
+++ b/tests/ui/typeck/issue-114529-illegal-break-with-value.stderr
@@ -1,0 +1,29 @@
+error[E0571]: `break` with value from a `while` loop
+  --> $DIR/issue-114529-illegal-break-with-value.rs:9:13
+   |
+LL |         while true {
+   |         ---------- you can't `break` with a value in a `while` loop
+LL |             break 9;
+   |             ^^^^^^^ can only break with a value inside `loop` or breakable block
+   |
+help: use `break` on its own without a value inside this `while` loop
+   |
+LL |             break;
+   |             ~~~~~
+
+error[E0571]: `break` with value from a `while` loop
+  --> $DIR/issue-114529-illegal-break-with-value.rs:16:13
+   |
+LL |         while let Some(v) = Some(9) {
+   |         --------------------------- you can't `break` with a value in a `while` loop
+LL |             break v;
+   |             ^^^^^^^ can only break with a value inside `loop` or breakable block
+   |
+help: use `break` on its own without a value inside this `while` loop
+   |
+LL |             break;
+   |             ~~~~~
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0571`.

